Source/carthage/Build.swift

```diff
@@ -20,74 +20,91 @@ public struct BuildCommand: CommandType {
 
 	public func run(mode: CommandMode) -> Result<(), CommandantError<CarthageError>> {
 		return producerWithOptions(BuildOptions.evaluate(mode))
-			|> map { options in
+			|> flatMap(.Merge) { options in
 				return self.buildWithOptions(options)
 					|> promoteErrors
 			}
-			|> flatten(.Merge)
 			|> waitOnCommand
 	}
 
 	/// Builds a project with the given options.
 	public func buildWithOptions(options: BuildOptions) -> SignalProducer<(), CarthageError> {
-		return self.createLoggingSink(options)
-			|> map { (stdoutSink, temporaryURL) -> SignalProducer<(), CarthageError> in
+		return self.openLoggingHandle(options)
+			|> flatMap(.Merge) { (stdoutHandle, temporaryURL) -> SignalProducer<(), CarthageError> in
 				let directoryURL = NSURL.fileURLWithPath(options.directoryPath, isDirectory: true)!
 
-				let (stdoutSignal, schemeSignals) = self.buildProjectInDirectoryURL(directoryURL, options: options)
+				var buildProgress = self.buildProjectInDirectoryURL(directoryURL, options: options)
+					|> flatten(.Concat)
+
+				let stderrHandle = NSFileHandle.fileHandleWithStandardError()
 
 				// Redirect any error-looking messages from stdout, because
 				// Xcode doesn't always forward them.
-				var grepDisposable: Disposable?
 				if !options.verbose {
-					let coldOutput = SignalProducer { observer, disposable in
-						disposable.addDisposable(stdoutSignal.observe(observer))
-					}
-
-					let stderrSink: FileSink<NSData> = FileSink.standardErrorSink()
-					stdoutSignal
-						|> filter { _ in false }
-						|> observe(stderrSink)
-
-					let stderrSinkWrapper: SinkOf<NSData> = SinkOf { data in
-						stderrSink.put(.Next(Box(data)))
-						return
-					}
-
-					let task = TaskDescription(launchPath: "/usr/bin/grep", arguments: [ "--extended-regexp", "(warning|error|failed):" ], standardInput: coldOutput)
-					grepDisposable = launchTask(task, standardOutput: stderrSinkWrapper).start()
+					let (stdoutProducer, stdoutSink) = SignalProducer<NSData, NoError>.buffer(0)
+					let grepTask: BuildSchemeProducer = launchTask(TaskDescription(launchPath: "/usr/bin/grep", arguments: [ "--extended-regexp", "(warning|error|failed):" ], standardInput: stdoutProducer))
+						|> on(next: { taskEvent in
+							switch taskEvent {
+							case let .StandardOutput(data):
+								stderrHandle.writeData(data)
+
+							default:
+								break
+							}
+						})
+						|> catch { _ in .empty }
+						|> then(.empty)
+						|> promoteErrors(CarthageError.self)
+
+					buildProgress = buildProgress
+						|> on(next: { taskEvent in
+							switch taskEvent {
+							case let .StandardOutput(data):
+								sendNext(stdoutSink, data)
+
+							default:
+								break
+							}
+						}, terminated: {
+							sendCompleted(stdoutSink)
+						}, interrupted: {
+							sendInterrupted(stdoutSink)
+						})
+
+					buildProgress = SignalProducer(values: [ grepTask, buildProgress ])
+						|> flatten(.Merge)
```

Comment:
Does `grepTask` termination be waited at somewhere?
Test with https://github.com/Carthage/ReactiveTask/commit/18cdc3fc44e65a1564bc3ff201d49230eae7a986 cause segmentation fault and seems `grepTask`'s threads are working.